common/include/RevCommon.h

```diff
@@ -59,32 +60,54 @@ constexpr T&& make_dependent( T&& x ) {
 
 /// Zero-extend value of bits size
 template<typename T>
-constexpr auto ZeroExt( T val, size_t bits ) {
-  return static_cast<std::make_unsigned_t<T>>( val ) & ( ( std::make_unsigned_t<T>( 1 ) << bits ) - 1 );
+constexpr auto ZeroExt( T val, int bits ) {
+  using UT = std::make_unsigned_t<T>;
+  return UT( val & ~( UT( ~UT{} ) << bits ) );
 }
 
 /// Sign-extend value of bits size
 template<typename T>
-constexpr auto SignExt( T val, size_t bits ) {
+constexpr auto SignExt( T val, int bits ) {
   auto signbit = std::make_unsigned_t<T>{ 1 } << ( bits - 1 );
-  return static_cast<std::make_signed_t<T>>( ( ZeroExt( val, bits ) ^ signbit ) - signbit );
+  return std::make_signed_t<T>( ( ZeroExt( val, bits ) ^ signbit ) - signbit );
 }
 
 /// Base-2 logarithm of integers
 template<typename T>
 constexpr int lg( T x ) {
   static_assert( std::is_integral_v<T> );
-
   // We select the __builtin_clz which takes integers no smaller than x
   if constexpr( sizeof( x ) <= sizeof( int ) ) {
-    return x ? 8 * sizeof( int ) - 1 - __builtin_clz( x ) : -1;
+    return x ? int{ 8 * sizeof( int ) - 1 } - __builtin_clz( x ) : INT_MIN;
```

Comment:
I don't understand. When `x==0` `lg() ` used to return `-1`, but now it returns `INT_MIN` (the closest integer to -Infinity).

Nothing has changed except the return value for `lg(0)`. `lg()` has existed for over a year.

Reply:
The `lg()`  function has always handled `0` as a special case. Look at the previous code. I just changed the code to explicitly cast the `size_t` value of `8 * sizeof (int ) -1` to `int{}` so that we don't get signedness-changing error, and I changed `-1` to `INT_MIN` for the return value of `lg(0)`.

I don't think we need to remind people that `lg(0) = -Infinity` or raise exceptions. This `lg()` function is `constexpr` and is intended for compile-time evaluation. It's not intended to be passed arguments from the end-user, but it passed arguments that we control in Rev. We are not going to call `lg()` with arguments dependent on end-user arguments that we need to flag as user errors.